### DISCUSSION TITLE
DialogueItem overwrite

### DISCUSSION THREAD

**Discussion by abeiro** [CharmedBaryon/CommonLibSSE-NG#67] [Q&A]
Hello,

Is it possible to overwrite the constructor of DialogueItem?

I noticed this code in the DialogueItem.h source file:

```
private:
		DialogueItem* Ctor(TESQuest* a_quest, TESTopic* a_topic, TESTopicInfo* a_topicInfo, Actor* a_speaker)
		{
			using func_t = decltype(&DialogueItem::Ctor);
			REL::Relocation<func_t> func{ RELOCATION_ID(34413, 35220) };
			return func(this, a_quest, a_topic, a_topicInfo, a_speaker);
		}
	};
```

so, commonlibsse-ng constructor is calling directly a function at this address : RELOCATION_ID(34413, 35220) . I suppose that's the real constructor.

It's is possible to overwrite that function at that address?, like a hook. Something like this:

```
namespace hook2 {

    class ResponseDataHook final : public RE::DialogueItem {
    public:
        RE::DialogueItem* CtorX(RE::TESQuest* a_quest, RE::TESTopic* a_topic, RE::TESTopicInfo* a_topicInfo,
                                RE::Actor* a_speaker);
        using func_t = decltype(&CtorX);
        inline static REL::Relocation<func_t> process;

        RE::DialogueItem* hookedCtor(RE::TESQuest* a_quest, RE::TESTopic* a_topic, RE::TESTopicInfo* a_topicInfo,
                                     RE::Actor* a_speaker) {
            return process(this, a_quest, a_topic, a_topicInfo, a_speaker);
        }
    };
    inline void install_hook() {


        REL::Relocation<std::uintptr_t> v_table_dm(RELOCATION_ID(34413, 35220).address());
        ResponseDataHook::process = v_table_dm.write_vfunc(0x9, &ResponseDataHook::hookedCtor);
    }

}
```

**Comment by jpstewart** [comment]
Yes, you could hook the constructor function at that address. Since CommonLibSSE itself is calling the original constructor it would not only intercept the call from the original engine but also any call from plugins that use CommonLibSSE.

However, in your example you are using the `write_vfunc` function. This only works on virtual functions in the vtable, which constructors never are. You will need to either hook the instruction that calls the instructor, or use something that can hook the callee (the function being called) such as Microsoft Detours and hook the function itself and not the vtable index.
